### DISCUSSION TITLE
The branch proto have interesting thing to talk about

### DISCUSSION THREAD

**Discussion by mackoj** [pointfreeco/swift-composable-architecture#832] [General]
Hi,

The work on the branch **proto** is interesting because it tackles multiple issues that I faced often when using TCA.

https://github.com/pointfreeco/swift-composable-architecture/compare/proto?expand=1

From what I understand `Environment` is removed and is replaced by a property wrapper `@Dependency`.

- this change how reducer work it ease how we can inject dependency into it
- sharing configuration through Environment is not ideal because it is often required in View and force us to duplicate these in the ViewState with this modification you can just call `@Dependency` in the view
- this removes all the custom Environment struct

Do you think that the work on this branch is interesting?

What other advantages/inconvenient that you see with this approach?

Does anyone tried to use this in his/her codebase ?

Thanks,

**Comment by iampatbrown** [comment]
@mackoj this branch is pretty awesome. Converting existing code to use `@Dependency` is straight forward and combined with a bit of `async/await` the overall ergonomics feels quite nice. 

I haven't really thought about the pros/cons of each approach and was probably too excited to really be objective when I was testing it out though :P

Interested to hear how other people's experiences are :)

**Comment by junebash** [comment]
This is something I was [experimenting with awhile back](https://forums.swift.org/t/closure-vs-function-performance-and-protocols/46776/2) as well, so I’m excited Brandon and Stephen have been looking into it more as well!

I’m especially interested in the ways dependencies are structured; it makes it even more of a parallel to SwiftUI. I think there may be an opportunity as well in having dependencies propagate down the Reducer hierarchy in a way that mirrors SwiftUI’s Environment propagation, though I’m not certain how this might be accomplished. I have some ideas, but I’ll need to sit down and play with them.

I was also experimenting with a couple of other ideas recently after browsing through this branch:
- `@resultBuilder` syntax for Reducers.
- in-line child reducer composition (eg including them on the reducer rather than having to compose them outside of the type or as a non-required static property).
- AsyncSequence effects; allowing the return of _any_ AsyncSequence (not `AnyAsyncSequence` 😉). This could almost be accomplished with another resultBulder, but it’s more difficult than SwiftUI’s ViewBuilder because of the associated Action type, meaning we can’t just return `some AsyncSequence`; we would have to have some theoretical Swift syntax that allows us to specify what the Element of the Effect would be, but as of now I don’t think it’s possible. 
- Related to the above, splitting out the mutation of state and the returning of Effects into different methods. This would make the theoretical EffectBuilder syntax easier to accomplish, as well as remove a whole lot of `return .none`. Unfortunately I’m not sure it’s be very efficient without even more rethinking, since in some cases you’d be having to switch over or even unwrap values twice (once in each method).

Ideally that'd all look something like:

```swift
protocol Reducer {
  associatedtype State
  associatedtype Action
  associatedtype Child: Reducer where Child.State == State, Child.Action == Action
  associatedtype Effect: AsyncSequence where Effect.Element == Action

  var child: Child { get }

  func apply(_ action: Action, to state: inout State)
  func fireEffects(for action: Action, with state: State) -> Effect
}

extension Reducer {
  func apply(_ action: Action, to state: inout State) {}
}

extension Reducer where Child == EmptyReducer<State, Action> {
  var child: Child { EmptyReducer() }
}

extension Reducer where Effect == EmptyAsyncSequence<Action> {
  func fireEffects(for action: Action, with state: State) -> Effect {
    EmptyAsyncSequence<Action>()
  }
}

struct AppReducer: Reducer {
  @ReducerBuilder<State, Action> // builder syntax will implicitly merge reducers
  var children: some Reducer<Action == AppAction, State == AppState> {
    SystemReducer()
      .pullback(/* ...*/)
      .optional()
    
    SettingsReducer(client: someExplicitDependency)
      .pullback(/* ... */)

    Feature1Reducer()
      .forEach(/* ... */)
      .dependency(/* ... */)
  }
  
  func apply(_ action: AppAction, to state: inout AppState) {
    switch action {
    case .doThing:
      state.thing = "thing"
      // etc...
    }
  }

  @AsyncSequenceBuilder<Action> // builder syntax implicitly creates conditional sequences
  func fireEffects(for action: AppAction, with state: AppState) -> some AsyncSequence<Element == Action> {
    switch action {
    case .didLaunch:
      self.someDependency.didLaunchAsyncSequence()
        .flatMap(/*...*/)
        .map(AppAction.blahblah)
    default:
      EmptyAsyncSequence()
    }
  }  
}
```

Again, a lot of that isn't yet possible, and I don't know how ideal a lot of it is. Some of it _is_ possible if one is will to, as we currently do, 'erase' the sequence to a universal, solid, generic `Effect` type, but this comes with some (probably mild in most cases) performance costs.

There’s a lot of fun to be had in this arena; I’m interested to see what, if anything, ends up sticking!

**Reply by stephencelis** [reply]
All interesting ideas! I agree that some of them seem to need more advanced compiler support, but if you end up trying anything interesting on our branch that works, we'd love to see it!

**Comment by stephencelis** [comment]
@mackoj Nice sleuthing! 😆

We're excited about the branch and what it's already unlocking, but it's definitely early days and there's a lot to figure out! The awkward interplay between top-level composed reducers (which we house in statics like `AppReducer.main` in the branch) vs. type-level reducers would ideally have a better story. @junebash's `Child` associated type (kinda like `Sequence.SubSequence`) is a particularly interesting way of potentially handling that. We also need to really kick the tires on it in a bigger project, like isowords, but it'll take some time to get there 😄

If anyone comes up with any interesting ideas or problems with the current branch, please do share!

**Comment by tgrapperon** [comment]
One concern I have with TCA is the huge memory traffic it can generate when composing and reducing states values in complex applications with deep nesting of features. I proposed something I believed would alleviate this in #603, but I was unaware of the copy-in/copy-out behavior of `inout` arguments in Swift, and this approach doesn't work.

One similar solution would be to pass to the reducer something wrapping the `State` instead of the raw `State`, in order to spot if reduction changes the state's value. This would allow bypassing re-assignation of unmodified states, decreasing the load on observers upstream. Because this would effectively change the signature of all reducers, I guess the `proto` branch offers an opportunity to explore the feasibility of this change, and what it would allow when combined with it. Please let me know if it's off-topic and another discussion should be opened instead.

Let define a `Snapshot<State>` struct:
```swift
@dynamicMemberLookup
public struct Snapshot<State> {
  public var state: State {
    didSet { stateDidChange = true }
  }
  var stateDidChange: Bool = false
  init(state: State) {
    self.state = state
  }
  
  public subscript<Value>(dynamicMember keyPath: KeyPath<State, Value>) -> Value {
    state[keyPath: keyPath]
  }
  
  public subscript<Value>(dynamicMember keyPath: WritableKeyPath<State, Value>) -> Value {
    get { state[keyPath: keyPath] }
    set { state[keyPath: keyPath] = newValue }
  }
}
```
This wrapper type is passed to the reducer:
```swift
public protocol _Reducer {
  associatedtype State
  associatedtype Action

  func reduce(into state: inout Snapshot<State>, action: Action) -> Effect<Action, Never>
}
```
Anything writing the state in `reduce` will raise the `stateDidChange` flag. If the value is false, we can forego state re-assignation after the reduction. Most of the accesses to `state` can be realized through `@dynamicMemberLookup`, and we still have the  `.state` field if we need to call `State`'s methods, for example.

As the `inout State` of the reducer tracks "internal" effects, and the `Effect<Action, Never>`, ("external") side effects, we can also imagine capturing both in the snapshot and define instead:
```swift
public struct Snapshot<State, Action> {
  public var state: State {
    didSet { stateDidChange = true }
  }
  public var effect: Effect<Action, Never>? = nil
  var stateDidChange: Bool = false

  init(state: State) {
    self.state = state
  }
}

public protocol _Reducer {
  associatedtype State
  associatedtype Action

  func reduce(snapshot: inout Snapshot<State, Action>, action: Action) -> Void
}
```
To produce an effect, we simply assign the snapshot's `.effect` property. We don't need to return an effect, and we could simply `break` of `noop` action cases. We can furthermore provide convenient syntactic sugar on `Snapshot` for `Effect(value:)` or to merge effects, for example. Some metadata can also be stored for debugging purposes, I guess.

Most importantly, this struct would capture the complete unit of work a reducer produces, and I have a feeling that this could be useful in async contexts, or when exploiting optimization opportunities the protocolized version of TCA would offer, for example.

What do you think about this?

**Reply by iampatbrown** [reply]
Interesting idea. Were you able to test this out at all? Could using `@dynamicMemberLookup` impact performance?  and are there advantages to doing this over a [copy on write wrapper](https://github.com/pointfreeco/swift-composable-architecture/discussions/488#discussioncomment-594289)?

If you get an implementation going, let me know. I'd be happy to check it out.

**Reply by tgrapperon** [reply]
Thanks for your feedback @iampatbrown. I don't have an implementation yet, as it requires changing code in a lot of places to get it building again, so I'm starting the discussion to gather other's ideas before diving-in.

`@dynamicMemberLookup` should indeed have a negative impact when assigning values (at best the same as keyPaths I'd say, maybe worse if it is fully dynamic at the end of the day (it comes from Swift for TensorFlow, to bind with Python API if I recall correctly), but I don't know for sure how it really works, and keyPaths versions are maybe optimized. I probably should look at it). 

This is however something that happens less frequently than state extraction/reassignation (which occurs for each emitted action, for each active reducer). The effect of `@dynamicMemberLookup` over direct access should be easy to benchmark, even in the current TCA implementation. I'll see if I can find the time to implement something this week.

The CoW wrapper is mostly ineffective for a child state, as this later is always reassigned to its parent at the end of its pulled-back reduction, invalidating the base value and creating a new copy in the parent.

**Reply by tgrapperon** [reply]
Using my [swift-measure](https://github.com/tgrapperon/swift-measure) library, I benchmarked `@dynamicMemberLookup` vs direct access and `KeyPath` for a wrapper type similar to `Snapshot`. I've tested the case where the wrapper is defined in an external module (like TCA), or in the same module, in order to check if whole-module-optimization was effective. I've learned a few interesting things of general interest along the way. The results are a little off topic too, so I'll post further developments in another discussion if needed.
```swift
@dynamicMemberLookup
public struct Snapshot<State> {
  public var state: State {
    didSet { stateDidChange = true }
  }

  public private(set) var stateDidChange: Bool = false
  public init(state: State) {
    self.state = state
  }

  public subscript<Value>(dynamicMember keyPath: WritableKeyPath<State, Value>) -> Value {
    get { state[keyPath: keyPath] }
    set { state[keyPath: keyPath] = newValue }
  }
}

struct State {
  var value: Int = 0
}
```
I've measured:
- Direct access: `snapshot.state.value`
- `@dynamicMemberLookup`: `snapshot.value`
- `KeyPath` on `State`: `snapshot.state[keyPath: \.value]`
- `KeyPath` on `Snapshot`: `snapshot[keyPath: \.state.value]`

Here are the results, on Xcode 13.1, in Release mode with `whole-module-optimization: on`:
```
== Snapshot Wrapper ==============================================================
                        Value  Difference Variation Performance  Error  Iterations
-- Snapshot State access ---------------------------------------------------------
Direct Access            73 ns                                  ±0.30 %    1000000
DynamicMemberLookup     342 ns    +281 ns +460.85 %       x0.18 ±0.09 %    1000000 (a)
KeyPath - Leaf      (*)  68 ns      -5 ns   -7.00 %       x1.08 ±0.33 %    1000000
KeyPath - Nested        349 ns    +277 ns +380.28 %       x0.21 ±0.09 %    1000000 (b)
-- Local Snapshot State access ---------------------------------------------------
Direct Access            44 ns                                  ±0.33 %    1000000
DynamicMemberLookup      43 ns      -1 ns   -1.85 %       x1.02 ±0.41 %    1000000
KeyPath - Leaf           43 ns      -0 ns   -0.77 %       x1.01 ±0.44 %    1000000
KeyPath - Nested    (*)  42 ns      -2 ns   -3.45 %       x1.04 ±0.33 %    1000000
==================================================================================
```
A few comments:
From the local (in-module) case, we learn that all access types are equivalents in terms of performance. This is a great! `@dynamicMemberLookup` doesn't seem to have any penalty cost at runtime.

The "external" (out-of-module) case is interesting too. All accesses are slower. I've managed to refactor and annotate the definition to retrieve an acceptable performance:
- The dynamic subscript needs to be annotated with `@_alwaysEmitIntoClient` to fix `(a)`.  `@inlinable` or `@inline(__always)` are ineffective.
- The `didSet` observation needs to be removed in order to fix `(b)`. One can use the subscript setter to raise the `stateDidChange` flag. We can't unfortunately protect the direct setter as `public private(set) var state: State` as this is incompatible with `@_alwaysEmitIntoClient`. Same with `public private(set) var stateDidChange` which need to be fully public too. Maybe there are annotations to prevent the user to directly set `state` and force them to pass through the subscript.
- 
Ultimately, if the `Snapshot` is defined as
```swift
@dynamicMemberLookup
public struct Snapshot<State> {
  public var state: State
  
  public var stateDidChange: Bool = false
  public init(state: State) {
    self.state = state
  }
  
  @_alwaysEmitIntoClient
  public subscript<Value>(dynamicMember keyPath: WritableKeyPath<State, Value>) -> Value {
    get { state[keyPath: keyPath] }
    set {
      state[keyPath: keyPath] = newValue
      stateDidChange = true
    }
  }
}
```
we finally get:
```
== Snapshot Wrapper =============================================================
                        Value Difference Variation Performance  Error  Iterations
-- Snapshot State access --------------------------------------------------------
Direct Access           60 ns                                  ±0.21 %    1000000
DynamicMemberLookup     59 ns      -1 ns   -1.38 %       x1.01 ±0.21 %    1000000 (a)
KeyPath - Leaf          58 ns      -2 ns   -3.08 %       x1.03 ±0.18 %    1000000
KeyPath - Nested    (*) 58 ns      -2 ns   -3.35 %       x1.03 ±0.17 %    1000000
-- Local Snapshot State access --------------------------------------------------
Direct Access           43 ns                                  ±0.22 %    1000000
DynamicMemberLookup     43 ns      +0 ns   +0.54 %       x0.99 ±0.34 %    1000000
KeyPath - Leaf          42 ns      -0 ns   -0.50 %       x1.01 ±0.26 %    1000000
KeyPath - Nested    (*) 42 ns      -1 ns   -1.45 %       x1.01 ±0.21 %    1000000
=================================================================================
```
Please note that only `(a)` is in a correct state after the mutations. In the three other cases, the `stateDidChange` wasn't raised, so this is not an acceptable solution.

I've also set up a few TCA benchmarks with a custom pullback using `Snapshot`'s to bypass unneeded re-assignations. The results are not really conclusive for now, as the time spent comparing or setting the states is a few orders of magnitude below the time spent in the whole TCA/Combine stack. We only gain 1% or 2% at best. It was a toy feature though, and it would be better to check with some real-life setup.
I'm trying to measure the effect on the stack size, but I can't get a good measure of it yet.

So, for now, there is nothing showing that avoiding unneeded re-assignations is a primary concern. There are maybe benefits in wrapping (state mutation & effect) in one type, but raw performance doesn't seem to be one of them, according to these preliminary results.

**Comment by mackoj** [comment]
Has someone tried to use [TaskLocal](https://developer.apple.com/documentation/swift/tasklocal) to store the environment ?

**Reply by mbrandonw** [reply]
Yes, absolutely! We have experimented with this on our parsing library, but haven't yet ported the ideas over to TCA. TLDR: it seems to work well and we have plans on using it once we have more time to experiment with it.

The only downside to use task locals is that there is a small performance price you pay when changing task local values. We don't think this price is going to be high for TCA, but it may require a delicate touch with our parsing library.

**Reply by mackoj** [reply]
Thanks for you response.

**Comment by andreyz** [comment]
Swift team in their [Big Picture](https://github.com/apple/swift-experimental-string-processing/blob/main/Documentation/BigPicture.md#event-processing) essay on string parsing, specifically talk about "event processing", that is using Swift's parsing abilities to handle and respond to events and how that generalises to application logic handling. Maybe someone will find it interesting here.

**Comment by JacksonUtsch** [comment]
Would be nice to remove `Environment`. Would adding `@Dependency` remove build time errors when dependencies are not injected like with `@EnvironmentObject`? Would be nice not to have to worry about this.

**Reply by tgrapperon** [reply]
The `@EnvironmentObject` comes from Vanilla SwiftUI and is not commonly used in TCA projects (only one is used internally when implementing `CaseLet` views). After a quick glance, this branch doesn't help in this area. Do you have a specific example in mind where some `@EnvironmentObject` interacts with a TCA stack?

**Reply by JacksonUtsch** [reply]
I referenced `@EnvironmentObject` because it seems like injecting a dependency with `@Dependency` has the same interface. `@Dependency` would require using the `.dependency` function on `_Reducer` as `@EnvironmentObject` uses the `.environmentObject` function.

`@EnvironmentObject` can interact with TCA just like `@ObservedObject` can with TCA, using a ViewStore. I've found that creating modular derived `Store`s allows for managing shared `State`

**Reply by tgrapperon** [reply]
Ah, sorry, I misread and didn't see the "*like* @EnvironmentObject".

Making sure that all `@Dependency` are fulfilled with the correct values at launch is definitely a problem with dependency management. The current version of TCA with explicit dependencies doesn't have this problem because you need to declare them in order to build. I'm maintaining an [accessory library](https://github.com/tgrapperon/swift-composable-environment) that manages TCA dependencies almost automatically, with coincidentally the same kind of API described in this branch. But it can't make sure at build time that the AppStore version of the app is getting the `.live` dependency for example. I think this branch has the same issue. There are things that can help (like defaulting dependencies values to "failing" variants). Maybe some stronger safeguards are possible.

However, once dependencies are correctly fulfilled at launch, passing them along the reducer chain can be handled automatically, and you basically "pullback" like you "scope", using state and action only.
